text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+`process::exit`) on "most Unix-like" operating systems, only the low 8
+bits of this value are significant.
+
+[std::process::exit]: https://doc.rust-lang.org/std/process/fn.exit.html
+
+## Changes to `main`
+[changes-to-main]: #changes-to-main
+
+From the perspective of the code that calls `main`, its signature is
+now generic:
+
+``` rust
+fn<T: Termination> main() -> T { ... }
+```
+
+It is critical that people _writing_ main should not have to treat it
+as a generic, though.  Existing code with `fn main() -> ()` should
+continue to compile, and code that wants to use the new feature should
+be able to write `fn main() -> TermT` for some concrete type `TermT`.
+
+I also don't know whether the code that calls `main` can accept a
+generic.  The `lang_start` glue currently receives an unsafe pointer
+to `main`, and expects it to have the signature `() -> ()`.  The
+abstractly correct new signature is `() -> Termination` but I suspect
+that this is currently not possible, pending `impl Trait` return types
+or something similar.  `() -> Box<Termination>` probably _is_ possible
+but requires a heap allocation, which is undesirable in no-std
+contexts.
+
+A solution to both problems is to implement the new semantics entirely
+within the compiler.  When it notices that `main` returns anything
+other than `()`, it renames the function and injects a shim:
+
+``` rust
+fn main() -> () {
+    let term = $real_main();
+    term.write_diagnostic(std::env::progname(), std::io::LOCAL_STDERR);
+    std::process::exit(term.exit_status());
+}
+```
+
+## `main` in nostd environments
+[main-in-nostd-environments]: #main-in-nostd-environments
+
+Some no-std environments do have a notion of processes that run and
+then exit, but they may or may not have notions of "exit status" or
+"error messages".  In this case, the signature of `main` should be
+unchanged, and the shim should simply ignore whichever aspects of
+`Termination` don't make sense in context.
+
+There are also environments where
+[returning from `main` constitutes a _bug_.][divergent-main] If you
+are implementing an operating system kernel, for instance, there may
+be nothing to return to.  Then you want it to be a compile-time error
+for `main` to return anything other than `!`.  If everything is
+implemented correctly, such environments should be able to get that
+effect by omitting all stock impls of `Termination` other than for
+`!`.  Perhaps there should also be a compiler hook that allows such
+environments to refuse to let you impl Termination yourself.
+
+[divergent-main]: https://internals.rust-lang.org/t/allowing-for-main-to-be-divergent-in-embedded-environments/4717
+
+## Test functions and doctests
+[test-functions-and-doctests]: #test-functions-and-doctests
+
+The harness for `#[test]` functions is very simple; I think it would
+be enough to just give `#[test]` functions the same shim that we give
+to `main`.  The programmer would be responsible for adjusting their
```

Comment:
> I think it would be enough to just give `#[test]` functions the same shim that we give to `main`.

This would not be compatible with `no_std` test runners like [utest](https://github.com/japaric/utest).

Supporting `#[test] fn() -> Result<_, _>` is actually simple. The `test` crate already has [an `enum`](https://github.com/rust-lang/rust/blob/2be750b2439aaaa760b95f5be5763e179ec0258b/src/libtest/lib.rs#L148) that represents `#[test] fn()`, `#[bench] fn(&mut Bencher)`, etc. We just have to add a new variant that represents the `Result` variant and lift the `rustc` restriction that functions with `#[test]` attribute must have signature `fn()`. Then it would be the job of [the test runner](https://github.com/rust-lang/rust/blob/2be750b2439aaaa760b95f5be5763e179ec0258b/src/libtest/lib.rs#L283) to handle the new case.

But that would be a different RFC than this one.